lib/Common/DataStructures/ClusterList.h

```diff
@@ -0,0 +1,410 @@
+    // note that it'll run on nodes that were never involved in any merges and are thus in solo sets
+    template<typename MapAccessoryType>
+    inline void Map(void(*callBack)(indexType, indexType, MapAccessoryType), MapAccessoryType accessory)
+    {
+        Assert(consolidated);
+        for (indexType i = 0; i < maxIndex; i++)
+        {
+            callBack(i, list[i], accessory);
+        }
+    }
+};
+
+/*
+ * SegmentClusterList differs from a normal ClusterList in that we allocate regions
+ * of the list instead of the entire list. This should provide good performance due
+ * to the multi-pass nature of the globopt; regions of code will generally generate
+ * stacksyms in one region per pass (since they're allocated incrementally), so the
+ * numbers used will normally be somewhat grouped. This saves us allocating most of
+ * the array in cases where we have large functions (which is more common than code
+ * input may indicate due to inlining), assuming that the data structure would need
+ * to be allocated on a per-block basis otherwise.
+ */
+
+template<typename indexType, class TAllocator, unsigned int numPerSegment = 16>
+class SegmentClusterList
+{
+private:
+    // The maximum index enabled; basically an array bounds for the set.
+    indexType maxIndex;
+    // The backing store for this ClusterList type is an array of pointers to subsegment
+    // structures, each of which stores a fixed region. Elements outside this region are
+    // assumed to have an identity mapping before first use.
+    indexType **backingStore;
+#if DBG
+    // During the merge process, we may end up with some situations where set membership
+    // is chained; that is, that node 7 is in set 5, but 5 is in set 4. Following all of
+    // the merge operations we have a step to reconcile all of this information and then
+    // update each element of the list to point to the least member of its set. This can
+    // be checked in debug builds to see if we've done this step or not.
+    bool consolidated = false;
+#endif
+    TAllocator* alloc;
+    inline void EnsureBaseSize(indexType inputMax)
+    {
+        // We want to ensure that we can store to inputMax, which means we need to grow to inputMax+1
+        indexType targetSize = inputMax + 1;
+        // round up to the next numPerSegment multiple
+        indexType max = (targetSize % numPerSegment == 0) ? targetSize : (targetSize + numPerSegment - (targetSize % numPerSegment));
+        // get segment counts for allocation
+        indexType newNumSegments = max / numPerSegment;
+        indexType curNumSegments = maxIndex / numPerSegment;
+        if (newNumSegments > curNumSegments)
+        {
+            // Note that if you want to use this with something other than an ArenaAllocator, then
+            // you'll have to add clean-up mechanics
+            indexType **tempList = AllocatorNewArray(TAllocator, alloc, indexType*, newNumSegments);
```

Comment:
The usual advice I've seen for this scenario is to ensure the array grows by at least some fixed ratio over the previous size, so that the cost of copying all of the existing data can be amortized to constant-per-insertion. Otherwise a large number of in-order additions could make it quadratic. I doubt the current usage is big enough or ordered enough to make this matter, but maybe it would help.

Reply:
While that's possible, we're likely not going to hit that case because of how we generally get clusters of symids in this structure. I can do some profiling later to see if there's a better growth strategy for our common cases.